src/storage/rect.rs

```diff
@@ -0,0 +1,632 @@
+///     .with_half_size(IVec2 { x: 8, y: 4 })
+///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+///     .build_default::<i32>();
+///
+/// assert_eq!(rect_map.get(Hex::new(0, 0)), Some(&0));
+/// ```
+///
+/// internally handle coordinate transform
+/// - `hex`
+/// - => `ij` offset coordinate
+/// - => `rc` 2D view of `Vec`
+/// - => `idx` `Vec` index
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "bevy_ecs",
+    derive(bevy_ecs::resource::Resource, bevy_ecs::component::Component)
+)]
+#[derive(Clone)]
+pub struct RectMap<T> {
+    inner: Vec<T>,
+    meta: RectMetadata,
+}
+
+/// Metadata struct for [`RectMap`]
+///
+/// # Example
+/// ```rust
+/// use hexx::{
+///     Hex, HexLayout, IVec2, OffsetHexMode,
+///     storage::{HexStore, RectMap, RectMetadata, WrapStrategy},
+/// };
+///
+/// let rect_hex_map = RectMetadata::default()
+///     .with_hex_layout(HexLayout::pointy().with_hex_size(1.0))
+///     .with_half_size(IVec2::new(8, 12))
+///     .with_offset_mode(OffsetHexMode::Odd)
+///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+///     .build_default::<i32>();
+/// let hex = Hex::new(0, 0);
+///
+/// assert_eq!(rect_hex_map.get(hex), Some(&0_i32));
+/// assert_eq!(rect_hex_map.wrapped_get(hex), &0_i32);
+/// ```
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "bevy_ecs",
+    derive(bevy_ecs::resource::Resource, bevy_ecs::component::Component)
+)]
+pub struct RectMetadata {
+    /// the hex layout of the map
+    hex_layout: HexLayout,
+    /// the offset mode of the map
+    ///
+    /// affect which way does the map zic zac.
+    offset_mode: OffsetHexMode,
+    /// the half size of the map
+    half_size: IVec2,
+    /// the wrapping strategy for indexing
+    wrap_strategies: [WrapStrategy; 2],
+}
+
+/// Wrapping Strategy for when try to [`RectMap::wrapped_get`] and
+/// [`RectMap::wrapped_get_mut`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+pub enum WrapStrategy {
+    /// Clamp (offset) coordinate to `min` and `max - 1`
+    ///
+    /// useful for latitude/vertical coordinate
+    Clamp,
+    /// Cycle (offset) coordinate to range of`min` and `max - 1`
+    ///
+    /// useful for longitude/horizontal coordinate
+    Cycle,
+}
+
+impl Default for RectMetadata {
+    fn default() -> Self {
+        Self {
+            hex_layout: HexLayout::pointy(),
+            offset_mode: OffsetHexMode::Odd,
+            half_size: IVec2::new(32, 16),
+            wrap_strategies: [WrapStrategy::Cycle, WrapStrategy::Clamp],
+        }
+    }
+}
```

Comment:
Having a default value here would be useful, but arbitrary (32, 16) values for the half size could create problems and confusion for users. I'd rather have no default impl at all.

Instead you could have a builder pattern like this:

```rust
RectMetadata::new(32, 16)
  .with_orientation(..)
  .with_offset_mode(...) ..etc;
```